src/librustc_mir/borrow_check.rs

```diff
@@ -1115,13 +1116,238 @@ enum NoMovePathFound {
+    Arbitrary,
+    AllOrNothing,
+    Disjoint,
+}
+
 impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
+    fn lvalue_element_conflict(&self,
+                               elem1: &Lvalue<'tcx>,
+                               elem2: &Lvalue<'tcx>)
+                               -> Overlap
+    {
+        match (elem1, elem2) {
+            (Lvalue::Local(l1), Lvalue::Local(l2)) => {
+                if l1 == l2 {
+                    debug!("lvalue_element_conflict: DISJOINT-OR-EQ-LOCAL");
+                    Overlap::AllOrNothing
+                } else {
+                    debug!("lvalue_element_conflict: DISJOINT-LOCAL");
+                    Overlap::Disjoint
+                }
+            }
+            (Lvalue::Static(_), _) |
+            (_, Lvalue::Static(_)) => {
+                // There's no point in tracking borrows to statics,
+                // so just ignore it.
+                debug!("lvalue_element_conflict: statics - ignore");
+                Overlap::Disjoint
+            }
+            (Lvalue::Projection(pi1), Lvalue::Projection(pi2)) => {
+                match (&pi1.elem, &pi2.elem) {
+                    (ProjectionElem::Deref, ProjectionElem::Deref) => {
+                        debug!("lvalue_element_conflict: DISJOINT-OR-EQ-DEREF");
+                        Overlap::AllOrNothing
+                    }
+                    (ProjectionElem::Field(f1, _), ProjectionElem::Field(f2, _)) => {
+                        if f1 == f2 {
+                            debug!("lvalue_element_conflict: DISJOINT-OR-EQ-FIELD");
+                            Overlap::AllOrNothing
+                        } else {
+                            let ty = pi1.base.ty(self.mir, self.tcx).to_ty(self.tcx);
+                            match ty.sty {
+                                ty::TyAdt(def, _) if def.is_union() => {
+                                    // We got stuck on a union.
+                                    debug!("lvalue_element_conflict: STUCK-UNION");
+                                    Overlap::Arbitrary
+                                }
+                                _ => {
+                                    debug!("lvalue_element_conflict: DISJOINT-FIELD");
+                                    Overlap::Disjoint
+                                }
+                            }
+                        }
+                    }
+                    (ProjectionElem::Downcast(_, v1), ProjectionElem::Downcast(_, v2)) => {
+                        // different variants are treated as having disjoint fields,
+                        // even if they occupy the same "space", because it's
+                        // impossible for 2 variants of the same enum to exist
+                        // (and therefore, to be borrowed) at the same time.
+                        //
+                        // Note that this is different from unions, but both
+                        // behaviors are grandfathered.
```

Comment:
> In any case, we could alter unions to behave similarly to enums if we wanted, I imagine, but not the opposite, right? Or is there a reason we can't?

I think we could make unions behave similarly to enums, it would just be "super-unsafe".